### PR TITLE
Feature messaging

### PR DESCRIPTION
**Pull request by sccolbert**

### CODE REVIEW

File: src/patterns/messaging.ts [hunk 1]
```diff
@@ -0,0 +1,556 @@
...
+   * given message is posted to the handler paired with this message.
+   * This message will already be enqueued and conflatable, and the
+   * given message will have the same `type` and also be conflatable.
+   *
+   * This method should merge the state of the other message into this
+   * message as needed so that when this message is finally delivered
+   * to the handler, it receives the most up-to-date information.
+   *
+   * If this method returns `true`, it signals that the other message
+   * was successfully conflated and it will not be enqueue.
```
Comment (dwillmer) [inline src/patterns/messaging.ts]: enqueued

Comment (sccolbert) [inline src/patterns/messaging.ts]: fixed

File: src/patterns/messaging.ts [hunk 2]
```diff
@@ -0,0 +1,556 @@
...
+        return;
+      }
+    }
+
+    // All message hooks returned true, so invoke the handler.
+    invokeHandler(handler, msg);
+  }
+
+  /**
+   * Post a message to handler for processing if the future.
```
Comment (dwillmer) [inline src/patterns/messaging.ts]: `a` handler .... `in` the future

File: src/patterns/messaging.ts [hunk 3]
```diff
@@ -0,0 +1,556 @@
...
+/**
+ * Install a message hook for a message handler.
+ *
+ * @param handler - The message handler of interest.
+ *
+ * @param hook - The message hook to install.
+ *
+ * #### Notes
+ * A message hook is invoked before a message is delivered to the
+ * handler. If the hook returns `false` from its [[hookMessage]]
```
Comment (blink1073) [inline src/patterns/messaging.ts]: should this be `[[processMessage]]`?

Comment (blink1073) [inline src/patterns/messaging.ts]: More like, this should be a `MessageHook` callback.

Comment (sccolbert) [inline src/patterns/messaging.ts]: yep thanks

**Pull request merged**